lms/djangoapps/verify_student/management/commands/send_verification_expiry_email.py

```diff
@@ -0,0 +1,174 @@
+    lies within the date range provided by command arguments. If the email is already sent indicated by field
+    `expiry_email_date` then resend after the specified number of days given as command argument `--resend_days`
+
+    The range to filter expired verification is selected based on --days-range. This represents the number of days
+    before now and gives us start_date of the range
+         Range:       start_date to today
+
+    The task is performed in batches with maximum number of users to send email given in `batch_size` and the
+    delay between batches is indicated by `sleep_time`.For each batch a celery task is initiated that sends the email
+
+    Example usage:
+        $ ./manage.py lms send_verification_expiry_email --resend-days=30 --batch-size=2000 --sleep-time=5
+    OR
+        $ ./manage.py lms send_verification_expiry_email
+
+    To run the command without sending emails:
+        $ ./manage.py lms send_verification_expiry_email --dry-run
+    """
+    help = 'Send email to users for which Software Secure Photo Verification has expired'
+
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '-d', '--resend-days',
+            type=int,
+            default=15,
+            help='Desired days after which the email will be resent to learners with expired verification'
+        )
+        parser.add_argument(
+            '--batch-size',
+            type=int,
+            default=1000,
+            help='Maximum number of users to send email in one celery task')
+        parser.add_argument(
+            '--sleep-time',
+            type=int,
+            default=10,
+            help='Sleep time in seconds between update of batches')
+        parser.add_argument(
+            '--days-range',
+            type=int,
+            default=365,
+            help="The number of days before now to check expired verification")
+        parser.add_argument(
+            '--dry-run',
+            action='store_true',
+            help='Gives the ranges of user_id for which email will be sent')
+
+    def handle(self, *args, **options):
+        """
+        Handler for the command
+
+        It creates batches of expired Software Secure Photo Verification against which email was not already sent
+        and initiates a celery task against it
+        """
+        resend_days = options['resend_days']
+        batch_size = options['batch_size']
+        sleep_time = options['sleep_time']
+        days = options['days_range']
+
+        start_date = datetime.now(UTC) - timedelta(days=days)
+        query = SoftwareSecurePhotoVerification.objects.filter(status='approved',
+                                                               expiry_date__lt=datetime.now(UTC),
+                                                               expiry_date__gt=start_date
+                                                               ).order_by('user_id')
+        sspv = use_read_replica_if_available(query)
```

Comment:
@zainab-amir  I would like to understand better why we cannot use edx-ace for this, it seems to be doing pretty much exactly the same thing we do for our nudge emails, upgrade reminders, and weekly highlights.

The patter we used for those emails was roughly this:

* Management command is called by jenkins
* It schedules N celery tasks which are each responsible for handling a chunk of users, we call this the "recipient resolver" since it's figuring out exactly who to send which emails to. We call a chunk of users a "bin".
* Each of those recipient resolver tasks is making one big SQL query to pull its bin of users. The bins are sized to ensure this data fits in memory. You can use different strategies for dividing up the data into bins, for example, you could assign each resolver task a bin number (0-N) and put a WHERE clause in your query that does something like user_id % number_of_bins == bin_num_for_this_task. Alternately you could divide them up based on hour of the day that they expired, giving you 24 bins.
* We then schedule a celery task for each row in that result set that actually sends the email.

It seems like you could follow exactly the same pattern here? The one thing you would need to do differently is also update the database after you send the email (like you are doing already below).

Happy to jump on a call to explain in more depth if needed!

Reply:
@mulby The decision to not use ace was taken before I started working on this command. One of the reasons was that we will need to make a separate network call to check if email was sent successfully. We need this information to update db field "expiry_email_date". 
This date will help us determine if enough time has passed and user has still not verified then we should send email again

Reply:
You don't actually have to use Sailthru, which I think may avoid these problems. Our password reset email, for example, uses ACE but sends the email using Django's email backend.

Using ACE allows the emails to be themed, translated etc.

At a minimum, you could replace your call to send_email to a call to edx_ace (much like we do for password reset), however, I suspect you could also leverage a bunch of the existing work we did on nudges etc to also streamline the fetching of the data etc.

Reply:
@mulby can you please give an example where this is used so that I can evaluate if it can be implemented for my scenario here ?

Reply:
@zainab-amir - thanks for taking this approach into consideration! The hope is that by standardizing our communication approach it will make it much easier to change our strategy in the future. For example, we may offer additional layers of user-controls and/or different messaging channels (like mobile notifications). Having all of the messages running through ACE will dramatically simplify those efforts.